src/main/webapp/3Dscript.js

```diff
@@ -0,0 +1,215 @@
+
+initThreeJs();
+gui();
+animate();
+
+/**
+ * Initializes the essential three.js 3D components then fetchs
+ * the pose data.
+ */
+function initThreeJs() {
+  scene = new THREE.Scene();
+  camera = new THREE.PerspectiveCamera(
+      /* fov =*/ 75,
+      /* aspectRatio =*/ window.innerWidth / window.innerHeight,
+      /* nearFrustum =*/.1,
+      /* farFrustum =*/ 1000);
+  camera.position.set(0, 1, 1);
+  renderer = new THREE.WebGLRenderer();
+  renderer.setSize(window.innerWidth, window.innerHeight);
+  document.body.appendChild(renderer.domElement);
+  clock = new THREE.Clock();
+  controls = new OrbitControls(camera, renderer.domElement );
+
+  fetchData();
+}
+
+
+/**
+ * This function fetchs pose data from the RunInfo servlet,
+ * it is an asynchronous call requiring addMap() to be
+ * called after the data is fully loaded.
+ */
+function fetchData() {
+  const queryString = window.location.search;
+  const urlParams = new URLSearchParams(queryString);
+  const id = urlParams.get('id');
+  const type = urlParams.get('dataType');
+  fetch('/getrun?id=' + id + '&dataType=' + type)
+      .then((response) => response.json())
+      .then((data) => pose = data)
+      .then(() => addMap());
+}
+
+/**
+ * Creates the 2D terrain and points a light at it.
+ * It also calls two other functions so it can wait on the fetch
+ * request and significantly cut down rendering time with
+ * geometry instancing.
+*/
+function addMap() {
+  // Creates the 2D map.
+  const loader = new THREE.TextureLoader();
+  let material = new THREE.MeshLambertMaterial({
+    map: loader.load(
+        'https://maps.googleapis.com/maps/api/staticmap' +
+      '?format=png&center=' + pose[0].lat + ',' + pose[0].lng +
+      '&zoom=18&size=500x500&key=' + apiKey),
+  });
+  let geometry = new THREE.PlaneGeometry(50, 50);
+  const map = new THREE.Mesh(geometry, material);
+  map.position.set(0, -4, 0);
+  map.rotation.x = -Math.PI / 2;
+  scene.add(map);
+
+  // Add the light to the scene.
+  const light = new THREE.PointLight(0xffffff, 1, 0 );
+  light.position.set(0, 100, 0);
+  scene.add(light);
+  plotTrajectory();
+
+  // Adds pose data to the scene as one Geometry instance.
+  material = new THREE.MeshBasicMaterial({color: 'red'});
+  geometry = new THREE.CylinderBufferGeometry(.005, .005, .1);
+  orientation = new THREE.InstancedMesh(geometry, material, pose.length);
+  orientation.instanceMatrix.setUsage(THREE.DynamicDrawUsage);
+  scene.add(orientation);
+  plotOrientation();
+}
+
+/**
+ * Converts lla coordinates to local world coordinates.
+ * @param {double} lat Latitude in degrees.
+ * @param {double} lng Longitude in degrees.
+ * @param {double} alt Altitude in meters.
+ * @return {array} An array returning gps coordinates.
+ */
+function llaDegreeToLocal(lat, lng, alt) {
+  lng = (lng - pose[0].lng) * 25000 * poseTransform.scale;
+  alt = (alt - pose[0].alt)/4;
+  lat = (lat - pose[0].lat) * 25000 * -poseTransform.scale;
+  return [lat, alt, lng];
```

Comment:
Technically these variables do not contain lat/lng/alt. Maybe `x, y, z` would make that clearer.